lifetimes/estimation.py

```diff
@@ -37,6 +38,117 @@ def _print_params(self):
+            np.abs(negative_log_likelihood_values) != np.inf]
+
+        penalizer_term = penalizer_coef * log(params).sum()
+        negative_log_likelihood = np.sum(negative_log_likelihood_values) + penalizer_term
+
+        if np.any(np.isnan(negative_log_likelihood_values)) or np.isnan(negative_log_likelihood):
+            return np.inf
+
+        return negative_log_likelihood
+
+    def conditional_expected_average_profit(self, x=None, m=None):
+        """
+        This method computes the conditional expectation of the average profit per transaction
+        for a group of one or more customers.
+            x: a vector containing the customers' frequencies. Defaults to the whole set of
+                frequencies used for fitting the model.
+            m: a vector containing the customers' monetary values. Defaults to the whole set of
+                monetary values used for fitting the model.
+
+        Returns:
+            the conditional expectation of the average profit per transaction
+        """
+        m = self.data['monetary_value'] if m is None else m
+        x = self.data['frequency'] if x is None else x
+        p, q, v = self.params_.values()
+        return np.mean((((q - 1) / (p * x + q - 1)) * (v * p / (q - 1))) + (p * x / (p * x + q - 1)) * m)
+
+    def customer_lifetime_value(self, transaction_prediction_model, frequency, recency, T, monetary_value, time=12, discount_rate=1):
+        """
+        This method computes the average lifetime value for a group of one or more customers.
+            transaction_prediction_model: the model to predict future transactions, literature uses
+                pareto/ndb but we can also use a different model like bg
+            frequency: the frequency vector of customers' purchases (denoted x in literature).
+            recency: the recency vector of customers' purchases (denoted t_x in literature).
+            T: the vector of customers' age (time since first purchase)
+            monetary_value: the monetary value vector of customer's purchases (denoted m in literature).
+            time: the lifetime expected for the user in months. Default: 12
+            discount_rate: the monthly adjusted discount rate. Default: 1
+
+        Returns:
+            the conditional expectation of the average profit per transaction.
+            Also creates a discounted_monthly_cash_flows attribute
+        """
+        df = DataFrame()
+        df['frequency'] = frequency
+        df['recency'] = recency
+        df['T'] = T
+        df['monetary_value'] = monetary_value
+
+        d = discount_rate
+        m = self.conditional_expected_average_profit()
+        self.discounted_monthly_cash_flows = []
+
+        for i in range(30, (time*30)+1, 30):
+            df['expected_revenues_period_'+str(i)] = df.apply(
+                lambda r: (m*transaction_prediction_model.predict(i, r['frequency'], r['recency'], r['T'])/(1+d)**(i/30)),
```

Comment:
not needed since i is range(30, (time*30)+1, 30) so basically [30, 60, 90...] third parameter in range defines the step, so i can always be integer divided by 30